experimental/slo/metrics.go

```diff
@@ -0,0 +1,227 @@
+	if !ok {
+		backend.Logger.Warn("Failed to sanitize datasource name for prometheus label", dsName)
+	}
+	return Metrics{DSName: dsName, DSType: dsType}
+}
+
+// WithEndpoint returns a new Metrics instance with the given endpoint
+func (m Metrics) WithEndpoint(endpoint Endpoint) Collector {
+	return Metrics{DSName: m.DSName, DSType: m.DSType, Endpoint: endpoint}
+}
+
+// CollectDuration collects the duration as a metric
+func (m Metrics) CollectDuration(source Source, status Status, statusCode int, duration float64) {
+	durationMetric.WithLabelValues(m.DSName, m.DSType, string(source), string(m.Endpoint), string(status), fmt.Sprint(statusCode)).Observe(duration)
+}
+
+// SanitizeLabelName removes all invalid chars from the label name.
+// If the label name is empty or contains only invalid chars, it will return false indicating it was not sanitized.
+// copied from https://github.com/grafana/grafana/blob/main/pkg/infra/metrics/metricutil/utils.go#L14
+func sanitizeLabelName(name string) (string, bool) {
+	if len(name) == 0 {
+		backend.Logger.Warn(fmt.Sprintf("label name cannot be empty: %s", name))
+		return "", false
+	}
+
+	out := strings.Builder{}
+	for i, b := range name {
+		if (b >= 'a' && b <= 'z') || (b >= 'A' && b <= 'Z') || b == '_' || (b >= '0' && b <= '9' && i > 0) {
+			out.WriteRune(b)
+		} else if b == ' ' {
+			out.WriteRune('_')
+		}
+	}
+
+	if out.Len() == 0 {
+		backend.Logger.Warn(fmt.Sprintf("label name only contains invalid chars: %q", name))
+		return "", false
+	}
+
+	return out.String(), true
+}
+
+// MetricsKey is a key for storing metrics in the context
+type MetricsKey string
+
+// DurationKey is a key for storing the duration in the context
+const DurationKey MetricsKey = "downstream_duration"
+
+// MetricsWrapper is a wrapper for a plugin that collects metrics
+type MetricsWrapper struct {
+	Name               string
+	Type               string
+	healthcheckHandler backend.CheckHealthHandler
+	queryDataHandler   backend.QueryDataHandler
+	resourceHandler    backend.CallResourceHandler
+	Metrics            Collector
+}
+
+// NewMetricsWrapper creates a new MetricsWrapper instance
+func NewMetricsWrapper(plugin any, s backend.DataSourceInstanceSettings, c ...Collector) *MetricsWrapper {
+	collector := NewMetrics(s.Name, s.Type)
+	if len(c) > 0 {
+		collector = c[0]
+	}
+	wrapper := &MetricsWrapper{
+		Name:    s.Name,
+		Type:    s.Type,
+		Metrics: collector,
+	}
+	if h, ok := plugin.(backend.CheckHealthHandler); ok {
+		wrapper.healthcheckHandler = h
+	}
+	if q, ok := plugin.(backend.QueryDataHandler); ok {
+		wrapper.queryDataHandler = q
+	}
+	if r, ok := plugin.(backend.CallResourceHandler); ok {
+		wrapper.resourceHandler = r
+	}
+	return wrapper
+}
+
+// QueryData calls the QueryDataHandler and collects metrics
+func (ds *MetricsWrapper) QueryData(ctx context.Context, req *backend.QueryDataRequest) (*backend.QueryDataResponse, error) {
+	ctx = context.WithValue(ctx, DurationKey, &Duration{Value: 0})
```

Comment:
Nit. Please use a different context key, see https://github.com/grafana/grafana-plugin-sdk-go/blob/e0452278f732f16376f36dc71c44cd107a13844b/backend/log/context.go#L5-L7 for example, to make sure the context key is unique. https://medium.com/random-go-tips/on-context-context-keys-uniqueness-738aa7a5c87f has a good explanation